### PR TITLE
Revert parts of #66405.

### LINKED ISSUE #67454: Increase in compilation times with newest beta & nightly
## Description

This was initially reported in https://github.com/Marwes/combine/issues/278.

I hit an important increase in compilation times on beta & nightly with a
particular configuration using `combine-3.8.1` since the recent version updates
in the Rust toolchains.

I tried building an example without `combine` in order to submit it to `rustc`
but it turns out not to be trivial. I ran into this as a side effect in CI and
can't spend much time investigating ATM, sorry.

From the CI logs I have at hand, the issue was not present as of:
rustc 1.41.0-nightly (fdc001156 2019-12-02).

## Reduced example

Here is an example with which I was able to reproduce the issue:

``` rust
use combine::byte::hex_digit;
use combine::{choice, token};
use combine::{ParseError, Parser, RangeStream};

fn parse<'a, I: 'a>() -> impl Parser<Input = I, Output = u8>
where
    I: RangeStream<Item = u8, Range = &'a [u8]>,
    I::Error: ParseError<I::Item, I::Range, I::Position>,
{
    choice!(
        token(b'A').map(|_| 1),
        token(b'B').map(|_| 2),
        token(b'C').map(|_| 3),
        token(b'D').map(|_| 4),
        token(b'E').map(|_| 5),
        token(b'F').map(|_| 6),
        token(b'G').map(|_| 7),
        token(b'H').map(|_| 8),
        token(b'I').map(|_| 9),
        token(b'J').map(|_| 10),
        token(b'K').map(|_| 11),
        token(b'L').map(|_| 12),
        token(b'M').map(|_| 13),
        token(b'N').map(|_| 14),
        token(b'O').map(|_| 15),
        token(b'P').map(|_| 16),
        token(b'Q').map(|_| 17),
//        token(b'R').map(|_| 18)
        (hex_digit(), hex_digit()).map(|(u, l)| {
            let val = (u << 4) | l;
            val
        })
    )
    .message("while parsing")
}

```

## Toolchain versions

The toolchain used for these tests are those available as of 2019/12/20:

- stable: rustc 1.40.0 (73528e339 2019-12-16)
- beta: rustc 1.41.0-beta.1 (eb3f7c2d3 2019-12-17)
- nightly: rustc 1.42.0-nightly (0de96d37f 2019-12-19)

## Compilation times

For each test, `cargo clean` was run before building in debug mode.

## With `message(..)`

| Configuration                      | stable | beta    | nightly |
| ---------------------------------- | ------ |-------- | ------- |
| 15 tokens                          | 12.79s |  26.83s |  25.38s |
| 17 tokens                          | 12.69s |  2m 20s |  2m 20s |
| 18 tokens                          | 12.89s |  6m 50s |  6m 55s |
| 17 tokens + (hex_digit, hex_digit) | 12.84s | 15m 57s | 16m 54s |

## Without `message(..)`

"Interestingly", removing the `.message(..)` line reduces compilation times
significantly.

| Configuration                      | stable | beta   | nightly |
| ---------------------------------- | ------ | ------ | ------- |
| 15 tokens                          | 12.67s | 19.89s |  16.14s |
| 18 tokens                          | 12.67s | 4m 11s |  2m 34s |
| 17 tokens + (hex_digit, hex_digit) | 12.78s | 5m 00s |  4m 35s |

### PR DESCRIPTION
**Pull request by nnethercote**
Because PR #66405 caused major performance regressions in some cases.

That PR had five commits, two of which affected performance, and three
of which were refactorings. This change undoes the performance-affecting
changes, while keeping the refactorings in place.

Fixes #67454.

r? @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Hmm, the perf results don't seem too reflect much improvement?

**Comment by Mark-Simulacrum** [comment]
We've had trouble with the perf results on this PR historically - the perf server is not quite accepting of the wins (or I guess losses too) that @nnethercote manages to measure locally. We're not sure why, though we've mostly eliminated bugs in the perf server itself such as measuring different artifacts.

It might be an architecture difference (e.g., @nnethercote is not using AMD chips) but that would be quite surprising to me.

**Comment by nnethercote** [comment]
This PR is expected to be a small performance loss on the benchmarks, because it reverts #66405, which was a small performance gain on the benchmarks, mostly for `inflate` and `keccak`.

But as @Mark-Simulacrum says, the perf results were very strange for that PR: inconsistent across multiple CI runs, all of which didn't really match my local results. (To the point where I did an experiment in #67248 to make sure the right revisions were being tested; they were.) And the results above are equally inconsistent, with as many wins as there are losses. And I have no idea why `unicode_normalization` has the largest change, it shouldn't be affected by `ObligationForest` changes. The whole thing is a mess.

Anyway, overall I think we should land this because it has minimal effect on the benchmark suite, but is known to fix some catastrophic performance regressions for real-world programs outside the benchmark suite (in #67454 and #67331). We should also backport this to beta, so that no rustc release suffers contains the regression.

I won't have a chance to analyze what went wrong in #66405 until I get back from PTO in late January, unfortunately.

One could also argue that the test program in #67454 should be added to the benchmark suite to prevent future regressions of this sort.

**Comment by sdroege** [comment]
> One could also argue that the test program in #67454 should be added to the benchmark suite to prevent future regressions of this sort.

Can code in the benchmarks depend on external crates? If so I can prepare a PR for this.

**Comment by nnethercote** [comment]
> Can code in the benchmarks depend on external crates? If so I can prepare a PR for this.

They can, and most of them do!

**Comment by sdroege** [comment]
Perfect, took me a while to find the right repo but here it is: https://github.com/rust-lang-nursery/rustc-perf/pull/588 :)

**Comment by sdroege** [comment]
What's blocking this to be merged?

**Comment by nikomatsakis** [comment]
Nothing, sorry, r=me -- but needs rebase

**Comment by memoryruins** [comment]
From https://github.com/rust-lang/rust/issues/67454#issuecomment-568068238
> I'm about to go on vacation for 6 weeks, so I will file a PR to back out #66405 before I go.

Since they return ~a day before the next release, someone else with permissions might need to handle the rebase.

**Comment by Mark-Simulacrum** [comment]
I will rebase and r=niko shortly.

**Comment by fengalin** [comment]
Many thanks, compilation times for [this use case](https://github.com/rust-lang/rust/issues/67454) are back to normal on nightly now.

Friendly reminder, @nnethercote [commented](https://github.com/rust-lang/rust/pull/67471#issuecomment-568214390):

> We should also backport this to beta, so that no rustc release contains the regression

File: src/librustc_data_structures/obligation_forest/mod.rs [hunk 1]
```diff
@@ -564,63 +550,51 @@ impl<O: ForestObligation> ObligationForest<O> {
...
-                        );
+        if node.state.get() == NodeState::Success {
+            match stack.iter().rposition(|&n| n == index) {
+                None => {
+                    stack.push(index);
+                    for &dep_index in node.dependents.iter() {
+                        self.find_cycles_from_node(stack, processor, dep_index);
                     }
+                    stack.pop();
+                    node.state.set(NodeState::Done);
```
Comment (Marwes) [inline src/librustc_data_structures/obligation_forest/mod.rs]: The slowdown is due to this line being lost, the added `dep_index >= min_index` check does not cover the case where a node points to a higher index. Which causes this to get `O(N ^ 2)` complexity.

**Comment by pnkfelix** [comment]
discussed at T-compiler meeting. Accepted for beta-backport.

File: src/librustc_data_structures/obligation_forest/mod.rs (hunk 1 above)
Comment (nnethercote) [reply src/librustc_data_structures/obligation_forest/mod.rs]: You are right, thanks for the info!

**Pull request merged**